qcodes/instrument_drivers/tektronix/AWG70000A.py

```diff
@@ -0,0 +1,936 @@
+            wfms: Binary .wfmx files. Should be packed like
+                [[wfmch1pos1, wfmch1pos2, ...], [wfmch2pos1, ...], ...]
+            seqname: The name of the sequence. This name will appear in the
+                sequence list. Note that all spaces are converted to '_'
+
+        Returns:
+            The binary .seqx file, ready to be sent to the instrument.
+        """
+
+        # input sanitising to avoid spaces in filenames
+        seqname = seqname.replace(' ', '_')
+
+        (chans, elms) = np.shape(wfms)
+        wfm_names = [['wfmch{}pos{}'.format(ch, el) for el in range(1, elms+1)]
+                     for ch in range(1, chans+1)]
+
+        flat_wfmxs = [wfmx for lst in wfms for wfmx in lst]
+        flat_wfm_names = [name for lst in wfm_names for name in lst]
+
+        sml_file = AWG70000A._makeSMLFile(trig_waits, nreps,
+                                          event_jumps, event_jump_to,
+                                          go_to, wfm_names, seqname)
+
+        user_file = b''
+        setup_file = AWG70000A._makeSetupFile(seqname)
+
+        buffer = io.BytesIO()
+
+        zipfile = zf.ZipFile(buffer, mode='a')
+        zipfile.writestr('Sequences/{}.sml'.format(seqname), sml_file)
+
+        for (name, wfile) in zip(flat_wfm_names, flat_wfmxs):
+            zipfile.writestr('Waveforms/{}.wfmx'.format(name), wfile)
+
+        zipfile.writestr('setup.xml', setup_file)
+        zipfile.writestr('userNotes.txt', user_file)
+        zipfile.close()
+
+        buffer.seek(0)
+        seqx = buffer.getvalue()
+        buffer.close()
+
+        return seqx
+
+    @staticmethod
+    def _makeSetupFile(sequence: str) -> str:
+        """
+        Make a setup.xml file.
+
+        Args:
+            sequence: The name of the main sequence
+
+        Returns:
+            The setup file as a string
+        """
+        head = ET.Element('RSAPersist')
+        head.set('version', '0.1')
+        temp_elem = ET.SubElement(head, 'Application')
+        temp_elem.text = 'Pascal'
+        temp_elem = ET.SubElement(head, 'MainSequence')
+        temp_elem.text = sequence
+        prodspec = ET.SubElement(head, 'ProductSpecific')
+        prodspec.set('name', 'AWG70002A')
+        temp_elem = ET.SubElement(prodspec, 'SerialNumber')
+        temp_elem.text = 'B020397'
+        temp_elem = ET.SubElement(prodspec, 'SoftwareVersion')
+        temp_elem.text = '5.3.0128.0'
+        temp_elem = ET.SubElement(prodspec, 'CreatorProperties')
+        temp_elem.set('name', '')
+
+        xmlstr = ET.tostring(head, encoding='unicode')
+        xmlstr = xmlstr.replace('><', '>\r\n<')
+
+        return xmlstr
+
+    @staticmethod
+    def _makeSMLFile(trig_waits: List[int],
+                     nreps: List[int],
+                     event_jumps: List[int],
+                     event_jump_to: List[int],
+                     go_to: List[int],
+                     wfm_names: List[List[str]],
```

Comment:
Are these modified in place otherwise I would in general suggest using Sequence see http://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance

Reply:
I'm not sure I understand the docs... The above code will fail if I pass something which is a `list` of a subclass of `int` because mypy considers `List` to be invariant? But immutable containers like `Sequence` are covariant. Is this why `Sequence` is preferred?

If so, I would say that it doesn't seem like such a big issue here, since we rarely subclass `int`, but I could be overlooking something crucial.

Reply:
So in general Sequence is preferable just to tell the user that it's imitable. The 
above would in principle be wrong if you pass a tuple as go_to